pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp

```diff
@@ -509,117 +513,68 @@ class Measurements final
         std::vector<std::size_t> all_offsets =
             Pennylane::Util::generateBitsPatterns(
-                Pennylane::Util::getIndicesAfterExclusion(sorted_wires,
-                                                          num_qubits),
+                Pennylane::Util::getIndicesAfterExclusion(wires, num_qubits),
                 num_qubits);
-
+        Kokkos::View<std::size_t *> d_all_offsets = vector2view(all_offsets);
         Kokkos::View<PrecisionT *> d_probabilities("d_probabilities",
                                                    all_indices.size());
-
-        Kokkos::View<std::size_t *> d_sorted_ind_wires("d_sorted_ind_wires",
-                                                       sorted_ind_wires.size());
-        Kokkos::View<std::size_t *> d_all_indices("d_all_indices",
-                                                  all_indices.size());
-        Kokkos::View<std::size_t *> d_all_offsets("d_all_offsets",
-                                                  all_offsets.size());
-
-        Kokkos::deep_copy(
-            d_all_indices,
-            UnmanagedSizeTHostView(all_indices.data(), all_indices.size()));
-        Kokkos::deep_copy(
-            d_all_offsets,
-            UnmanagedSizeTHostView(all_offsets.data(), all_offsets.size()));
-        Kokkos::deep_copy(d_sorted_ind_wires,
-                          UnmanagedSizeTHostView(sorted_ind_wires.data(),
-                                                 sorted_ind_wires.size()));
-
-        const int num_all_indices =
-            all_indices.size(); // int is required by Kokkos::MDRangePolicy
-        const int num_all_offsets = all_offsets.size();
-
-        MDPolicyType_2D mdpolicy_2d0({{0, 0}},
-                                     {{num_all_indices, num_all_offsets}});
-
-        Kokkos::parallel_for(
-            "Set_Prob", mdpolicy_2d0,
-            KOKKOS_LAMBDA(const std::size_t i, const std::size_t j) {
-                const std::size_t index = d_all_indices(i) + d_all_offsets(j);
-                const PrecisionT REAL = arr_data(index).real();
-                const PrecisionT IMAG = arr_data(index).imag();
-                const PrecisionT value = REAL * REAL + IMAG * IMAG;
-                Kokkos::atomic_add(&d_probabilities(i), value);
-            });
-
-        std::vector<PrecisionT> probabilities(all_indices.size(), 0);
-
-        if (is_sorted_wires) {
-            Kokkos::deep_copy(UnmanagedPrecisionHostView(probabilities.data(),
-                                                         probabilities.size()),
-                              d_probabilities);
-            return probabilities;
+        Kokkos::View<ComplexT *> sv = this->_statevector.getView();
+
+        // Reducing over `d_probabilities` requires too much L0 scratch memory
+        // on GPUs. If n_wires >= 20, this also requires quite a bit of memory
+        // on CPUs, so we fallback to the next implementation
+        if (n_wires < 20 && !is_gpu_scratch_limited) {
```

Comment:
This is faster, but requires quite a bit of memory as `d_probabilities`'s data may be duplicated across threads.